### PR TITLE
docs/advanced: add initial draft of data-by-pass

### PR DESCRIPTION
**Pull request by w4tsn**
Hello there! I found the time to write up a first draft for the data by-pass pattern mentioned in #545 on our approach.

I've tried to comply roughly around the idea of pattern documents, feel free to comment on that decision.

Some things I want to point out where I think discussion is required:

* Under Discussion I've added first points to benefits, drawbacks and issues that came to my mind. Please add / comment on that as you see fit. Maybe there are actual issues that can be mapped for reference to this doc
* The image does not yet fit into the design of the docs in general, happy for feedback / additions here
* Currently this docs page has no references to the rest of the docs although that might be a good idea tying things together. Any ideas?
* I've added a new section Advanced concepts right under Basic concepts, what do you think?
* The data-by-pass pattern makes a pre-assumption on the pattern to augment things with features from additional services through the messaging API, that may need it's own document
* Currently the Advanced section has no further structure / hierarchy, any objections?

I hope that little christmas present is to your liking and you can actually appreciate it / it's not too rough on the edges :)

### CODE REVIEW

**Comment by w4tsn** [comment]
@thjaeckle in case you missed it, I'd like to hear your opinion on this! And welcome back ;)

**Comment by thjaeckle** [comment]
@w4tsn thx, I'll have a look at it soon - probably end of this week.

File: documentation/src/main/resources/pages/ditto/advanced-data-by-pass.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+In order for the client-application to retrieve the requested data in a secure and performant way we introduce a high-performance proxy (e.g. based on nginx, example below). The proxy will not have any credentials by itself, it's just serving prepared queries on a randomly, hard-to-guess URL with an expiration time of 5 minutes. It features an admin API which the micro-service has credentials to access.
+
+The provider micro-service hooks into a twin (e.g. via websockets) and listens for queries. If a query arrives it will formulate the query, store it at the high-performance proxy (which might already query the data) and return a randomly generated URL to the proxy together with a Location-header as a response to the client-application. The client then needs to follow the response in order to retrieve the data from the proxy.
+
+With this approach the access to the database is secured via ditto policies and in scope of single things while the data retrieval happens via a performant proxy application without the ditto cluster ever seeing those packages.
```
Comment (thjaeckle) [inline documentation/src/main/resources/pages/ditto/advanced-data-by-pass.md:36]:
```suggestion
With this approach the access to the database is secured via Ditto [policies](basic-policy.html) and in scope of single things while the data retrieval happens via a performant proxy application without the Ditto cluster ever seeing those packages.
```

File: documentation/src/main/resources/pages/ditto/advanced-data-by-pass.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+## Discussion
+
+**Benefits**:
+
+* Higher performance compared to using just ditto
```
Comment (thjaeckle) [inline documentation/src/main/resources/pages/ditto/advanced-data-by-pass.md:42]:
```suggestion
* Higher performance compared to using just Ditto
```

File: documentation/src/main/resources/pages/ditto/advanced-data-by-pass.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+* The ditto policy system can be utilized to scope and secure data access from clients to databases / -stores
```
Comment (thjaeckle) [inline documentation/src/main/resources/pages/ditto/advanced-data-by-pass.md:43]:
```suggestion
* The Ditto [policy system](basic-policy.html) can be utilized to scope and secure data access from clients to databases / -stores
```

File: documentation/src/main/resources/pages/ditto/advanced-data-by-pass.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+**Drawbacks**:
+
+* A third-party application for the high-performance proxy has to be added and maintained
+* A custom messaging API is necessary in the first place introducing a higher complexity
+* A translation of certain query-languages from messages to the actual databases / applications has to be implemented
+
+**Issues**:
+
+* Managing and communicating custom messaging APIs is not natively supported in ditto, other ways have to be explored to keep APIs consistent
```
Comment (thjaeckle) [inline documentation/src/main/resources/pages/ditto/advanced-data-by-pass.md:53]:
```suggestion
* Managing and communicating custom messaging APIs is not natively supported in Ditto, other ways have to be explored to keep APIs consistent
```

File: documentation/src/main/resources/pages/ditto/advanced-data-by-pass.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+The ceryx proxy is a modified nginx with a redis-database to store the randomly generated IDs correlating with prepared queries. It is not suited for this use-case on it's own so capabilities to store queries (including Authentication) behind expiring random URLs was added, but not send upstream yet.
+
+## Known uses
+
+**[othermo GmbH](https://www.othermo.de) uses this for a history-service**: The history service connects to ditto via websockets and hooks into things by answering specific `/history` messages. The messages API is translated to InfluxDB queries which then are stored with a randomly generated URL and expiration of 5 minutes at the high-performance proxy. The service then returns the random URL to the client which then follows the 303 to retrieve the actual data.
```
Comment (thjaeckle) [inline documentation/src/main/resources/pages/ditto/advanced-data-by-pass.md:63]:
```suggestion
**[othermo GmbH](https://www.othermo.de) uses this for a history-service**: The history service connects to Ditto via websockets and hooks into things by answering specific `/history` messages. The messages API is translated to InfluxDB queries which then are stored with a randomly generated URL and expiration of 5 minutes at the high-performance proxy. The service then returns the random URL to the client which then follows the 303 to retrieve the actual data.
```

**Review by thjaeckle — CHANGES_REQUESTED** [review]
Hi @w4tsn and thanks for this first added documentation of advanced patterns, looks very good already 👍 
I added some suggestions (e.g. like writing Ditto always with capital "D" ;)) and mainly have a remark to the added image in combination with the text.

One more thing for the new added "tag" is missing: Please adjust the following file and add the new tag, otherwise it won't show at the bottom of the page:
https://github.com/eclipse/ditto/blob/master/documentation/src/main/resources/_data/tags.yml

File: documentation/src/main/resources/pages/ditto/advanced-data-by-pass.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+The solution is comprised of the following systems:
+
+* **database**: where your bigger chunks of data reside and wait to be delivered / queried
+* **database provider mirco-service**: the service managing the database connection and exposing it to clients through things messaging API
+* **thing**: a digital thing with extended API through a micro-service
+* **client**: a client-application trying to receiving bigger quantities of data via a things messaging API in the scope of that thing and secured via ditto policies
+* **high-performance data proxy** (or just proxy): a third-party application proxy sitting in-between the database and the provider micro-service managing data delivery
+
+{% include image.html file="pages/advanced/data-by-pass-architectural-design.jpg" alt="Architectural Design" caption="Architectural design of the data by-pass pattern showing all actors and their interactions." max-width=800 %}
```
[the remark this answers is not in the document]
Comment (w4tsn) [reply documentation/src/main/resources/pages/ditto/advanced-data-by-pass.md:30]: Oh yeah, that was sloppy. I'll update the graphic to match the more generalized theme.

[the remark this answers is not in the document]
Comment (w4tsn) [reply documentation/src/main/resources/pages/ditto/advanced-data-by-pass.md:30]: I've generalized the terms used in the diagram. Please have a look if that's acceptable.

**Comment by w4tsn** [comment]
@thjaeckle I've added another section on the 'recommended' policy design with this pattern.

**Review by thjaeckle — APPROVED** [review]

File: documentation/src/main/resources/pages/ditto/advanced-data-by-pass.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (thjaeckle) [reply documentation/src/main/resources/pages/ditto/advanced-data-by-pass.md:30]: @w4tsn looks good, thx 👍

**Comment by w4tsn** [comment]
@thjaeckle this is ready from my end. Thanks a lot for your review(s)!

**Pull request merged**